### ISSUE TITLE
Comment on "Finding Your Data With Views"

### ISSUE THREAD

**Issue by janl** [oreilly/couchdb-guide#75]
This summary introduces many new ideas which were not covered in the
chapter, and is thus NOT at all a summary. Each of the topics
discussed in the reduce summary should be dealt with in detail within
the chapter.

I think reduce requires a slower and more detailed description with
real-world examples taken step by step. Perhaps a small reduce problem
with five documents would be appropriate for this introductory
section.

From http://books.couchdb.org/relax/design-documents/views

> _Reduce functions_ operate on the sorted rows emitted by map view functio
> ns. CouchDB=92s reduce functionality takes advantage of one of the fundamen=
> tal properties of B-tree indexes: for every leaf node (a sorted row), there=
>  is a chain of internal nodes reaching back to the root. Each leaf node in =
> the B-tree carries a few rows (on the order of tens, depending on row size)=
> , and each internal node may link to a few leaf nodes or other internal nod=
> es.

**Comment by aaronvegh** [comment]
Agreed, and totally disappointed that this issue is over a year old and untouched. This chapter is a mind-boggling example of ignorance about who is reading this thing -- I had never seen an example of a Reduce function prior to this chapter; the dependence on understanding B-Trees was surprising, I have no idea what "logarithmic time" is about. Please, get someone else to re-write this chapter, because it's doing the whole book a disservice.

**Comment by janl** [comment]
@aaronvegh I appreciate the input, but we all have a lot to do. If you are so frustrated, why not propose a better version of the paragraphs in question? I'm happy to look over a pull request and commit it.

**Comment by aaronvegh** [comment]
It's good to get a response to this thread; it's discouraging to see an original complaint -- and a valid one go unanswered for over a year.

The problems with the chapter are that it drops the reader unceremoniously into topics without any kind of warning or introduction. We got a gentle introduction to Map functions earlier, but a very rude intro to Reduce functions without explanation for what they are for. The rest of the book does a terrific job of taking the reader through the topics, building on previous knowledge. That practice is abandoned in this chapter.

The problems begin with the heading "The View to Get Comments for Posts". The opening paragraph sets the tone, and doesn't recover: 

> "We use an array key here to support the group_level reduce query parameter. CouchDB’s views are stored in the B-tree file structure (which will be described in more detail later on). Because of the way B-trees are structured, we can cache the intermediate reduce results in the non-leaf nodes of the tree, so reduce queries can be computed along arbitrary key ranges in logarithmic time."

If you read that as someone who is new to CouchDB, a lot of that isn't going to make sense. We've heard of reduce functions, but know nothing about the query parameter. We keep hearing about B-Tree structures, but they don't really make any sense. Therefore, we don't know about leaf vs. non-leaf nodes of the tree. And since we're not compsci grads, the notion of logarithmic time is daunting, if not downright obtuse.

I ended up going to the Apache CouchDB Wiki, which has a chapter with the same topic -- http://wiki.apache.org/couchdb/Introduction_to_CouchDB_views I'm sure you're familiar with it. 

I also found this interactive tutorial for understanding MapReduce, which was also very useful: http://blog.mudynamics.com/wp-content/uploads/2009/04/icouch.html

While I have increased my understanding somewhat, I couldn't say I've mastered it sufficiently to write an accurate chapter on it. 

Overall, this chapter felt like it was written by a different author. The following chapter happily resumes the pace, and I hope this is just one aberration. 

Thanks,
Aaron
